File: Python/Dynamic-Programming/Min_Jump.py
#Name :  Atul Kumar
#Github username : atul1510
#Repositary name : Algorithms
# Problem Statement: C++ program to find find the minimum number of jumps to reach the other end of the array starting from the 0th index. And if the end isn’t reachable, return -1 as the output.
#Example:
#1] INPUT: int arr[] = { 1, 9, 4, 7, 3, 7, 2, 1, 8, 9, 5 };
#OUTPUT: Minimum number of jumps to reach the other end is 2 */
# Python3 program to find Minimum number of jumps to reach end
# Returns minimum number of jumps to reach arr[n-1] from arr[0]
def minJump(arr, x):
	jump = [0 for i in range(x)]
	if (x == 0) or (arr[0] == 0):
		return float('inf')
	jump[0] = 0
	# To Find the minimum number of jumps to reach arr[i] from
	for i in range(1, x):
		jump[i] = float('inf')
		for j in range(i):
			if (i <= j + arr[j]) and (jump[j] != float('inf')):   # arr[0] and assign this value to jumps[i]
				jump[i] = min(jump[i], jump[j] + 1)
				break
	return jump[x-1]
# Main Function
arr = [ 1, 9, 4, 7, 3, 7, 2, 1, 8, 9, 5 ]
size = len(arr)
print('Minimum number of jumps to reach',
	'end is', minJump(arr, size))

File: Java/Dynamic-Programming/Min_Jump.java
/*Name :  Atul Kumar
Github username : atul1510
Repositary name : Algorithms
// Problem Statement: C++ program to find find the minimum number of jumps to reach the other end of the array starting from the 0th index. And if the end isn’t reachable, return -1 as the output.
//Example:
1] INPUT: int arr[] = { 1, 9, 4, 7, 3, 7, 2, 1, 8, 9, 5 };
OUTPUT: Minimum number of jumps to reach the other end is 2 */
// JAVA Code for Minimum number of jumps to reach end
import java.io.*;
public class Min_Jump {
	private static int minJump(int[] arr, int n)
	{   // jump[n-1] will contain the result
		int jump[] = new int[n];
		int i, j;
		// if first element is 0,
		if (n == 0 || arr[0] == 0)
			return Integer.MAX_VALUE;
		// end cannot be reached
		jump[0] = 0;
        // Find the minimum number of jumps to reach arr[i]  from arr[0], and assign the value to jump[i]
		for (i = 1; i < n; i++) {
			jump[i] = Integer.MAX_VALUE;
			for (j = 0; j < i; j++) {
				if (i <= j + arr[j]
					&& jump[j] != Integer.MAX_VALUE) {
					jump[i]
						= Math.min(jump[i], jump[j] + 1);
					break;
				}
			}
		}
		return jump[n - 1];
	}
	// Main function
	public static void main(String[] args)
	{
		int array[] = { 1, 9, 4, 7, 3, 7, 2, 1, 8, 9, 5 };
		System.out.println(
			"Minimum number of jumps to reach end is : "
			+ minJump(array, array.length));
	}
}